File: day7/solution2.pl
#!/usr/bin/env perl

use strict;
use warnings;
use v5.14;

use constant TOTAL_SPACE      => 70_000_000;
use constant MIN_UPDATE_SPACE => 30_000_000;

sub dir_size ($\%) {
  my ($dir, $sizes) = @_;

  my $size = 0;
  for my $key (keys %$sizes) {
    if ($key =~ /^$dir\//) {
      $size += $sizes->{$key};
    }
  }

  return $size;
}

open my $fh, '<', 'input.txt' or die $!;

my @path = ("");
my %size = ("" => 0);

while (my $line = <$fh>) {
  chomp $line;

  if ($line =~ /^\$ cd \//) {
    @path = ("");
  }
  elsif ($line =~ /^\$ cd \.\./) {
    pop @path;
  }
  elsif (my ($chdir) = $line =~ /^\$ cd (.+)$/) {
    push @path, $chdir;
  }
  elsif (my ($lsdir) = $line =~ /^dir (.+)$/) {
    my $dir_path = join '/', @path, $lsdir;
    $size{$dir_path} = 0;
  }
  elsif (my ($size, $file) = $line =~ /^(\d+) (.+)$/) {
    my $file_path = join '/', @path, $file;
    $size{$file_path} = $size;
  }
}

my $used = dir_size '', %size;
my $free = TOTAL_SPACE - $used;
my $need = MIN_UPDATE_SPACE - $free;
my $min;

for my $path (keys %size) {
  next unless $size{$path} == 0;

  my $size = dir_size $path, %size;

  next if $size < $need;
  next if defined $min && $size > $min;

  $min = $size;
}

say $min;

File: day7/solution2.exs
#!/usr/bin/env elixir

defmodule TerminalLogReader do
  defstruct files: %{}, path: []

  @total_space 70_000_000
  @min_for_update 30_000_000

  def new(files \\ %{"" => 0}, path \\ []) do
    %__MODULE__{
      files: files,
      path: path
    }
  end

  def scan(input_file) do
    File.stream!(input_file)
    |> Enum.reduce(new(), &scan_line/2)
  end

  def scan_line(line, acc) do
    line
    |> String.trim()
    |> String.split()
    |> parse(acc)
  end

  def parse(["$", "cd", "/"], acc), do: new(acc.files, [])
  def parse(["$", "cd", ".."], acc), do: new(acc.files, cdr(acc.path))
  def parse(["$", "cd", node], acc), do: new(acc.files, [node | acc.path])
  def parse(["$", "ls"], acc), do: acc
  def parse(["dir", name], acc), do: new(acc.files |> Map.put(mkpath(acc, name), 0), acc.path)

  def parse([str_size, name], acc) do
    {size, _rem} = Integer.parse(str_size)

    new(
      acc.files |> Map.put(mkpath(acc, name), size),
      acc.path
    )
  end

  defp cdr([_ | cdr]), do: cdr
  defp mkpath(acc, file), do: [file | acc.path] |> Enum.reverse() |> Enum.join("/")

  def list_dir_sizes(acc) do
    acc.files
    |> Map.keys()
    |> Enum.filter(fn d -> acc.files[d] == 0 end)
    |> Enum.map(fn dir ->
      {
        dir,
        acc.files
        |> Map.filter(fn {file, _} -> String.starts_with?(file, dir) end)
        |> Map.values()
        |> Enum.sum()
      }
    end)
    |> Map.new()
  end

  def collect_sizes(acc) do
    sizes = acc |> list_dir_sizes()
    used = sizes[""]
    free = @total_space - used
    need = @min_for_update - free

    sizes
    |> Map.filter(fn {_, size} -> size >= need end)
    |> Map.values()
    |> Enum.min()
  end
end

TerminalLogReader.scan("input.txt")
|> TerminalLogReader.collect_sizes()
|> IO.puts()
